### PR TITLE
Added capsule mesh

### PR DESCRIPTION
**Pull request by ahcorde**
Related with this issue https://github.com/ignitionrobotics/ign-common/issues/140

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

File: graphics/include/ignition/common/MeshManager.hh [hunk 1]
```diff
@@ -154,6 +154,18 @@ namespace ignition
...
                                   const int _segments);
 
+     /// \brief Create a capsule mesh
+     /// \param[in] _name the name of the new mesh
+     /// \param[in] _radius the radius of the capsule in the x y plane
+     /// \param[in] _length length of the body
+     /// \param[in] _rings the number of circles along the height
+     /// \param[in] _segments the number of segment per circle
+     public: void CreateCapsule(const std::string &_name,
+                                const float radius,
```
Comment (mjcarroll) [inline graphics/include/ignition/common/MeshManager.hh:164]: Same float/double questions as #154

File: graphics/src/MeshManager.cc [hunk 2]
```diff
@@ -822,6 +822,160 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
 
+void MeshManager::CreateCapsule(const std::string &_name,
+                                const float _radius,
+                                const float _length,
+                                const unsigned int _rings,
+                                const unsigned int _segments)
+{
+  if (this->HasMesh(_name))
+  {
+    return;
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:833]: Add test for this, same as #154

File: graphics/src/MeshManager.cc [hunk 3]
```diff
@@ -822,6 +822,160 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
+  }
+
+  Mesh *mesh = new Mesh();
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:836]: Not for here, but I'm going to open an issue about moving away from raw pointers in MeshManagerPrivate.

File: graphics/src/MeshManager.cc [hunk 4]
```diff
@@ -822,6 +822,160 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
+  mesh->SetName(_name);
+  this->dataPtr->meshes.insert(std::make_pair(_name, mesh));
+
+  ignition::math::Vector3d vert, norm;
+  SubMesh subMesh;
+
+  // Based on https://github.com/godotengine/godot primitive_meshes.cpp
+  int prevRow, thisRow, point;
+  float x, y, z, u, v, w;
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:845]: float vs double

File: graphics/src/MeshManager.cc [hunk 5]
```diff
@@ -822,6 +822,160 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
+  point = 0;
+
+  /* top hemisphere */
+  thisRow = 0;
+  prevRow = 0;
+  for (unsigned int j = 0; j <= (_rings + 1); j++) {
+    v = j;
+
+    v /= (_rings + 1);
+    w = sin(0.5 * M_PI * v);
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:858]: Replace M_PI

File: graphics/src/MeshManager.cc [hunk 6]
```diff
@@ -822,6 +822,160 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
+  if (this->HasMesh(_name))
+  {
+    return;
+  }
+
+  Mesh *mesh = new Mesh();
+  mesh->SetName(_name);
+  this->dataPtr->meshes.insert(std::make_pair(_name, mesh));
+
+  ignition::math::Vector3d vert, norm;
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:840]: vert and norm are unused in this impl

**Review by mjcarroll — COMMENTED** [review]
Largely the same comments as #154

File: graphics/src/MeshManager.cc [hunk 7]
```diff
@@ -822,6 +822,159 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
+      u /= _segments;
+
+      x = -sin(u * (IGN_PI * 2.0));
+      z = cos(u * (IGN_PI * 2.0));
+
+      ignition::math::Vector3d p(
+      x * _radius * w, y, -z * _radius * w);
+      // Compute vertex
+      subMesh.AddVertex(ignition::math::Vector3d(
+      p + ignition::math::Vector3d(0.0, 0.5 * _length, 0.0)));
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:871]: Indent this continuation.

File: graphics/src/MeshManager.cc [hunk 8]
```diff
@@ -822,6 +822,159 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
+      subMesh.AddTexCoord(ignition::math::Vector2d(u, v * oneThird));
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:872]:
```suggestion
      subMesh.AddTexCoord({u, v * oneThird});
```

File: graphics/src/MeshManager.cc [hunk 9]
```diff
@@ -822,6 +822,159 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
+      x = -sin(u * (IGN_PI * 2.0));
+      z = cos(u * (IGN_PI * 2.0));
+
+      ignition::math::Vector3d p(
+      x * _radius, y, -z * _radius);
+
+      // Compute vertex
+      subMesh.AddVertex(p);
+      subMesh.AddTexCoord(
+      ignition::math::Vector2d(u, oneThird + (v * oneThird)));
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:914]:
```suggestion
      subMesh.AddTexCoord({u, oneThird + (v * oneThird)});
```

File: graphics/src/MeshManager.cc [hunk 10]
```diff
@@ -822,6 +822,159 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
+      u2 /= _segments;
+
+      x = -sin(u2 * (IGN_PI * 2.0));
+      z = cos(u2 * (IGN_PI * 2.0));
+
+      ignition::math::Vector3d p(
+      x * _radius * w, y, -z * _radius * w);
+      // Compute vertex
+      subMesh.AddVertex(ignition::math::Vector3d(
+      p + ignition::math::Vector3d(0.0, -0.5 * _length, 0.0)));
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:954]: indent

File: graphics/src/MeshManager.cc [hunk 11]
```diff
@@ -822,6 +822,159 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
+      subMesh.AddTexCoord(
+      ignition::math::Vector2d(u2, twoThirds + ((v - 1.0) * oneThird)));
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:956]:
```suggestion
      subMesh.AddTexCoord({u2, twoThirds + ((v - 1.0) * oneThird)});
```

**Review by mjcarroll — COMMENTED** [review]
Just a few style nits

File: graphics/src/MeshManager.cc [hunk 12]
```diff
@@ -822,6 +822,157 @@ void MeshManager::CreateCamera(const std::string &_name, float _scale)
...
+  for (unsigned int j = 0; j <= (_rings + 1); j++) {
+    v = j;
+
+    v /= (_rings + 1);
+    v += 1.0;
+    w = sin(0.5 * IGN_PI * v);
+    y = _radius * cos(0.5 * IGN_PI * v);
+
+    for (unsigned int i = 0; i <= _segments; i++) {
+      float u2 = i;
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:943]: To make Windows happy

```suggestion
      double u2 = static_cast<double>(i);
```

**Review by mjcarroll — APPROVED** [review]
LGTM once windows comes back happy.

Thanks for iterating so quickly!

**Comment by ahcorde** [comment]
These two failing tests are not part of this PR

 - VideoEncoderTest.Exists

**Pull request merged**